### ISSUE TITLE
Find library stack to enable MQTT over WSS

### ISSUE THREAD

**Issue by stelfrag** [netdata/netdata#7570]
###### background info (#7413)
- The libmosquitto library could not be used to upgrade the connection to wss
- Attempted to switch and use the libpaho library
  - Was able to use mqtt over websockets
  - Was able to use mqtt over SSL
  - Was not to use mqtt over secure websockets (seems like an issue with the libpaho C library)

##### Next steps
- [x] Post a comment on the libpaho project page asking about an issue that's being worked on and may resolve the failure to upgrade to secure websockets
- [x] Focus again on the libmosquitto -- post a Q on the project for assistance (if there is a way to upgrade to wss)

##### Expected behavior
The agent must be able to use the library to establish a connection to a, websockets enabled MQTT broker, over SSL and upgrade to websockets (i.e. wss)

**Comment by stelfrag** [comment]
Added an issue with a [question](https://github.com/eclipse/mosquitto/issues/1540) on the eclipse/mosquitto project about the libmosquitto library to see if there is a way to upgrade a connection to wss

Added a [comment](https://github.com/eclipse/paho.mqtt.c/issues/721) on the eclipse/paho.mqtt.c project asking for more information on a reported bug (that seems to be related with the issue we are dealing) and whether we could use the feedback to help provide a fix if posssible

**Comment by amoss** [comment]
Reading https://mosquitto.org/ChangeLog.txt indicates that minor releases are 1-2 months apart. Shorter when it is mostly bug-fixes in the library code, and longer for new features.

Running through their open issues to try to estimate which will impact our use-case:
* https://github.com/eclipse/mosquitto/issues/1526
* https://github.com/eclipse/mosquitto/issues/1527 (we are not running on a mac but strange behaviour in the underlying websockets library could cause us problems)
* https://github.com/eclipse/mosquitto/issues/1507
* https://github.com/eclipse/mosquitto/issues/1506 (potential incompatability with our far end)
* https://github.com/eclipse/mosquitto/issues/1397
* https://github.com/eclipse/mosquitto/issues/1358 (could affect our use-case)
* https://github.com/eclipse/mosquitto/issues/1289 (VerneMQ incompatability would be a problem for us)
* https://github.com/eclipse/mosquitto/issues/942 (although he is working in C++ this problem could affect our use)
* https://github.com/eclipse/mosquitto/issues/450 (needs to be checked if it affects us)
* https://github.com/eclipse/mosquitto/issues/406 (seems highly relevant)
* https://github.com/eclipse/mosquitto/issues/362 (we might need this)

What I would like to understand is how we can estimate the following:
* How much effort is initially required to get us to a working ACKL using 1. a patched libmosquitto, 2. a forked copy of the mosquitto repo, 3. directly ripping out the parts that we need and building them into the agent directly.
* How much ongoing maintenance effort is required for those same three options?
* How do these options compare to any other options that we have (e.g. libpaho or searching for another option)?

**Comment by stelfrag** [comment]
- 1526 seems to be fixed (tcp no delay option)
- 1527 refers to the broker (we will use the library)
  - It could be related to libwebsockets how it integrates with the broker. We'll see if that affects the library as well when support is added (assuming that the library will use libwebsockets as well which is highly likely)
- 1507 I found no reason to get the socket fd and mess with it directly.
- 1506 Again this refers to the broker (mosquitto). We know from the paho.mqtt.c for example that currectly it cannot do wss on a mosquitto broker (python lib does). I assume the javascript has similar problems as the C version (in addition to some breaking change that is mentioned in the issue).
- 1397 It compiles as described in the issue (WITH_WEBSOCKETS) has no effect in the client library as the libmosquitto doesn't support websockets. However, studying the code a bit, reveals that it changes the behavior of some functions of the lib/ sources when compiling them to generate the broker.  
- 1358 The library supports SOCK5 ( haven't tested it). It could affect us, I don't know in which scenario though. 
- 1259 This refers to broker <-> broker communication, not client to vernemq. It could potentially affects us as the libmosquitto shares code with the broker. Client to vernemq however is a different case as it will not involved the bridging features. 
- 942 I don't think it will affect us
- 450 It is thread safe (except for the initial init function). I found that it works fine (more extensive checks will be done of course) if we start our own thread (as per netdata model) and call `mosquitto_loop` every X seconds to handle operations)
- 406 Again this refers to the broker. When supported is added to the library we will make sure of course it doesn't segfault as it will kill netdata in that case (not good) :-) 
- 362 I don't think this has initial impact on us (tuning at the ssl layer). However we can investigate if tuning that value will increase performance (maybe faster transfer from client to broker at the ssl layer)


> How much effort is initially required to get us to a working ACKL using 1. a patched libmosquitto, 2. a forked copy of the mosquitto repo, 3. directly ripping out the parts that we need and building them into the agent directly.

1. Patching libmosquitto means we need to study 
    - libwebsockets
    - Study the mosquitto (broker) code in particular the code that accepts and upgrades the connection.
    - Add the upgrade connection request  (maybe using an appropriate libwebsockets api call -- I haven't studied that) via a new libmosquitto api call
1. We will need a fork to add the new functionality and:
    - Either keep it internally and maintain it
    - Assuming we do a fix / PR it will go to the official repo (and take time for the fix to reach all distros etc)
    - We need to check if we want to actually have library dependency or have it buildin
1. We would need to keep the lib/ part + libwebsockets dependency + the fix we did in part 2. This will introduce a libwebsockets dependency not libmosquitto. Not keeping the mosquitto fork in this case it means we will need check / address / fix issues ourselves (if they affect the code we took)

**Comment by stelfrag** [comment]
Focusing on studying the mosquitto code to see how we can add the functionality we want

**Comment by stelfrag** [comment]
- The mosquitto repository (v1.6.8)  has been forked under netdata
- When cloned inside netdata ,it can be compiled to produce a static library libmosquitto.a
   - The library resides under mosquitto/lib/
   - Include file (libmosquitto.h) is under mosquitto/lib as well
- Changes to the agent compile process were made to ensure the library is statically linked
- Changes to support ws and wss will be made to forked copy of the library

The higher level ACLK work proceeds in parallel

**Comment by amoss** [comment]
We can close this issue after the spike issue is created (so that we can put a link to the new one to allow people to track what is happening).

**Comment by stelfrag** [comment]
Closing this in favor of #7739
